Make auth callback state bounded

aherrmann-da opened the pull request:
Closes #7721 

- Uses a caffeine LRU cache as bounded storage rather than an unbounded `TrieMap` for pending login requests on the middleware and pending authorized requests on the trigger service.
- Adds tests that earlier requests are dropped when the bound is exceeded on both the auth middleware itself and the middleware client (i.e. trigger service).
- Makes these bounds configurable via CLI options on the auth middleware and trigger service.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

garyverhaegen-da commented:
/azp run

garyverhaegen-da commented:
/azp run

garyverhaegen-da commented:
/azp run

cocreature commented on triggers/service/auth/src/main/scala/com/daml/auth/middleware/api/Client.scala:
```diff
@@ -24,15 +24,18 @@ import akka.http.scaladsl.server.Directives._
 import akka.http.scaladsl.unmarshalling.Unmarshal
 import com.daml.auth.middleware.api.Client.{AuthException, RefreshException}
 import com.daml.auth.middleware.api.Tagged.RefreshToken
+import com.github.benmanes.caffeine.cache.{Cache, Caffeine}
```
Can we use `com.daml.caching` instead which wraps the caffeine cache?

aherrmann-da replied:
`com.daml.caching.Cache` does not expose `invalidate`, which we need here.

cocreature replied:
Makes sense, maybe worth considering adding it to `com.daml.caching` at some point but not in this PR.

cocreature reviewed (commented):
Nice, I have one question on the design here. If I understand this correctly, a later request will evict a previous one even if the timeout for that request has not yet expired. Is that correct? If so, is that intentional? To me it seems more intuitive to reject the later request with a 503 until the first has expired. Otherwise you can spam the auth middleware with requests and if you do it fast enough nothing will ever succeed.

aherrmann-da commented:
> Nice, I have one question on the design here. If I understand this correctly, a later request will evict a previous one even if the timeout for that request has not yet expired. Is that correct? If so, is that intentional? To me it seems more intuitive to reject the later request with a 503 until the first has expired. Otherwise you can spam the auth middleware with requests and if you do it fast enough nothing will ever succeed.

Correct, that's a good point. Responding with 503 seems like the better option. I've changed the code accordingly. In this case Caffeine's LFU cache no longer seems appropriate. I've added a small class based on `LinkedHashMap` that implements what we need here: It does not add new elements when the maximum capacity is reached and it evicts elements after a timeout.

cocreature commented on triggers/service/auth/src/main/scala/com/daml/auth/middleware/api/RequestStore.scala:
```diff
@@ -0,0 +1,77 @@
+    now - timestamp >= timeout.toNanos
+  }
+
+  private def evictTimedOut(now: Long): Unit = {
```
Hm, the fact that this is linear in the number of items seems a bit annoying but probably fine at least for now.

aherrmann-da replied:
It's linear in the number of timed-out items. Items are stored in increasing order of timestamp and this uses `.takeWhile` to find the prefix of timed out items.

cocreature commented on triggers/service/auth/src/main/scala/com/daml/auth/middleware/api/RequestStore.scala:
```diff
@@ -0,0 +1,77 @@
+    * Remove and return the value under the given key, if present and not timed out.
+    */
+  def pop(key: K): Option[V] = {
+    synchronized {
```
This will still return potentially timed out elements. Is that intentional? I guess there is no harm in having a longer timeout for a few elements.

aherrmann-da replied:
How so? There's a check whether the retrieved item is timed out and if so `pop` returns `None`.

cocreature replied:
duh, sorry I missed that check.

cocreature commented on triggers/service/auth/src/main/scala/com/daml/auth/middleware/oauth2/Config.scala:
```diff
@@ -24,15 +29,21 @@ case class Config(
 )
 
 object Config {
+  val DefaultMaxLoginRequests: Int = 10000
```
seems a bit high maybe 1000 or even 100?

aherrmann-da replied:
done

cocreature reviewed (approved):
Nice! Thanks for addressing my suggestion.
